contracts/libraries/NFTDescriptor.sol

```diff
@@ -0,0 +1,305 @@
+import '@uniswap/v3-core/contracts/libraries/BitMath.sol';
+import '@uniswap/v3-core/contracts/libraries/FullMath.sol';
+import '@openzeppelin/contracts/utils/Strings.sol';
+import '@openzeppelin/contracts/math/SafeMath.sol';
+import '@openzeppelin/contracts/math/SignedSafeMath.sol';
+import './HexStrings.sol';
+
+library NFTDescriptor {
+    using TickMath for int24;
+    using Strings for uint256;
+    using SafeMath for uint256;
+    using SafeMath for uint160;
+    using SafeMath for uint8;
+    using SignedSafeMath for int256;
+    using HexStrings for uint256;
+
+    struct ConstructTokenURIParams {
+        address token1;
+        address token0;
+        string token0Symbol;
+        string token1Symbol;
+        uint8 token0Decimals;
+        uint8 token1Decimals;
+        int24 tickLower;
+        int24 tickUpper;
+        int24 tickSpacing;
+        uint24 fee;
+        uint256 liquidity;
+        address poolAddress;
+    }
+
+    // TODO: use SafeERC20Namer, submit final draft of description text
+    function constructTokenURI(ConstructTokenURIParams memory params) internal pure returns (string memory) {
+        string memory name =
+            string(
+                abi.encodePacked(
+                    'Uniswap V3 - ',
+                    feeToPercentString(params.fee),
+                    ' - ',
+                    params.token0Symbol,
+                    '/',
+                    params.token1Symbol,
+                    ' - ',
+                    tickToDecimalString(
+                        params.tickLower,
+                        params.tickSpacing,
+                        params.token0Decimals,
+                        params.token1Decimals
+                    ),
+                    '<>',
+                    tickToDecimalString(
+                        params.tickUpper,
+                        params.tickSpacing,
+                        params.token0Decimals,
+                        params.token1Decimals
+                    )
+                )
+            );
+        string memory description =
+            string(
+                abi.encodePacked(
+                    'Represents a liquidity position in a Uniswap V3 pool. Redeemable for owed reserve tokens.',
+                    '\\nliquidity: ',
+                    uint256(params.liquidity).toString(),
+                    '\\npoolAddress: ',
+                    addressToString(params.poolAddress),
+                    '\\ntoken0Address: ',
+                    addressToString(params.token0),
+                    '\\ntoken1Address: ',
+                    addressToString(params.token1)
+                )
+            );
+
+        return
+            string(abi.encodePacked('data:application/json,{"name":"', name, '", "description":"', description, '"}'));
+    }
+
+    struct DecimalStringParams {
+        // significant figures of decimal
+        uint256 sigfigs;
+        // length of decimal string
+        uint8 bufferLength;
```

Comment:
yah... I really wanted to try and minimize the amount of params....not that this necessarily justifies it but zeppelin requires it in their `toHexString` function (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/243adff49ce1700e0ecb99fe522fb16cff1d1ddc/contracts/utils/Strings.sol#L55)

Getting rid of it would mean I'd have to do a redundant iteration over the sigfigs to get how many there are...unless there's a better way to get the "length" of a number???